## Importing Library

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.axes as ax

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import KFold
from sklearn import datasets

In [3]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Loading Data

In [4]:
data = datasets.load_boston()
data

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

#### Data

In [37]:
x=data.data
x_temp = pd.DataFrame(x)
x_temp.head(10)
x_temp

0     1      2    3      4      5     6       7    8      9     10  \
0    0.00632  18.0   2.31  0.0  0.538  6.575  65.2  4.0900  1.0  296.0  15.3   
1    0.02731   0.0   7.07  0.0  0.469  6.421  78.9  4.9671  2.0  242.0  17.8   
2    0.02729   0.0   7.07  0.0  0.469  7.185  61.1  4.9671  2.0  242.0  17.8   
3    0.03237   0.0   2.18  0.0  0.458  6.998  45.8  6.0622  3.0  222.0  18.7   
4    0.06905   0.0   2.18  0.0  0.458  7.147  54.2  6.0622  3.0  222.0  18.7   
..       ...   ...    ...  ...    ...    ...   ...     ...  ...    ...   ...   
501  0.06263   0.0  11.93  0.0  0.573  6.593  69.1  2.4786  1.0  273.0  21.0   
502  0.04527   0.0  11.93  0.0  0.573  6.120  76.7  2.2875  1.0  273.0  21.0   
503  0.06076   0.0  11.93  0.0  0.573  6.976  91.0  2.1675  1.0  273.0  21.0   
504  0.10959   0.0  11.93  0.0  0.573  6.794  89.3  2.3889  1.0  273.0  21.0   
505  0.04741   0.0  11.93  0.0  0.573  6.030  80.8  2.5050  1.0  273.0  21.0   

         11    12  
0    396.90  4.98  
1    396.90  9.14  
2    392.83  4.03  
3    394.63  2.94  
4    396.90  5.33  
..      ...   ...  
501  391.99  9.67  
502  396.90  9.08  
503  396.90  5.64  
504  393.45  6.48  
505  396.90  7.88  

[506 rows x 13 columns]

#### Output values

In [6]:
y=data.target
y
y_temp = pd.DataFrame(y)
y_temp.head(10)

0
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2
5  28.7
6  22.9
7  27.1
8  16.5
9  18.9

#### Data Features

In [7]:
features = data.feature_names
features

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

## Normal Data Splitting

In [8]:
xtrain, xtest, ytrain, ytest = train_test_split(x,y, test_size= 0.2)

In [9]:
xtrain.shape

(404, 13)

<font color = 'red'> Here we can see we have only 404 data to train our model which is very less

<font color = 'red'> passing this data to ML model without K-Fold

# Random Forest Regression (1)

In [10]:
rf = RandomForestRegressor(100)
rf = rf.fit(xtrain,ytrain)

#### predictions

In [11]:
predictions = rf.predict(xtest)
predictions

array([22.104, 20.959, 20.726,  9.199, 21.197, 15.886, 18.638, 24.455,
       20.356, 32.464, 25.011, 17.327, 16.741, 16.886, 17.762, 23.64 ,
        9.091, 16.23 , 14.665, 18.071, 16.837, 19.628, 32.622, 22.874,
        7.979, 22.556, 15.586, 18.274, 20.311, 25.758, 11.059, 22.949,
       39.493, 23.544, 19.539,  9.062, 33.81 , 47.858, 25.984, 25.05 ,
       23.656, 27.393, 14.215, 31.699, 19.475, 20.804, 15.016, 21.789,
       19.175, 19.375, 19.603, 20.888, 23.64 , 32.659, 24.573, 19.735,
       10.848, 19.824, 13.459, 20.327, 46.471, 44.326, 20.564, 20.357,
       20.424, 15.703, 15.684, 23.931, 19.803, 28.843, 10.867, 15.132,
       20.356, 11.453, 20.62 , 18.376,  9.394, 20.213, 11.389, 14.278,
       21.294, 21.183, 21.264, 22.63 , 12.749, 21.455, 34.092, 23.238,
       20.7  , 23.249, 20.784, 13.756, 22.145, 14.264, 25.295, 13.911,
       25.14 , 29.287, 19.693, 19.562,  9.433, 34.715])

#### Error or Accuracy

In [12]:
error = mean_squared_error(ytest,predictions)
error

10.818335401960786

# <font color = 'green'> K-Fold Cross Validation

In [32]:
kfold = KFold(n_splits = 5 , shuffle = False)

for train_index, test_index in kfold.split(x):
    
    print("Train :\n",train_index,'\nTest: \n',test_index)
    print("==================\n\n")

Train :
 [102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119
 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155
 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173
 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191
 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209
 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227
 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245
 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260 261 262 263
 264 265 266 267 268 269 270 271 272 273 274 275 276 277 278 279 280 281
 282 283 284 285 286 287 288 289 290 291 292 293 294 295 296 297 298 299
 300 301 302 303 304 305 306 307 308 309 310 311 312 313 314 315 316 317
 318 319 320 321 322 323 324 325 326 327 328 329 330 331 332 333 334 335
 336 337 338 339 340 341 342 343 344 345 3

# Random Forest Regression (2) 

<font color = 'red'> passing new K-fold data in this

In [14]:
rf1 = RandomForestRegressor()

error = []

for train_index, test_index in kfold.split(x):
    
    xtrain, xtest = x[train_index] , x[test_index]
    ytrain , ytest = y[train_index] , y[test_index]
    
    
    rf1=RandomForestRegressor()
    rf1 = rf1.fit(xtrain,ytrain)
    
    # Predictions
    
    prediction = rf1.predict(xtest)
    
    #error
    error.append(mean_squared_error(ytest,prediction))

In [15]:
error

[8.543275549019594,
 12.968922603960413,
 21.145991702970257,
 46.810352158415824,
 18.554323524752466]

<font color = 'red'> Here we can see model is mugging up the data instead of learning the reult the error is thismuch variable 

In [19]:
type(x)

numpy.ndarray